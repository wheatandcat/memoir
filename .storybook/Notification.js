import * as Notifications from 'expo-notifications';
import React, { memo, useContext, useCallback, createContext } from 'react';

export const Context = createContext({});
const { Provider } = Context;

Notifications.setNotificationHandler({
  handleNotification: async () => ({
    shouldShowAlert: true,
    shouldPlaySound: false,
    shouldSetBadge: false,
  }),
});

const Notification = memo((props) => {
  const onPermissionRequest = useCallback(async () => {
    return true;
  }, []);

  return <Provider value={{ onPermissionRequest }}>{props.children}</Provider>;
});

export default Notification;

export const useNotification = () => useContext(Context);
export const Consumer = Context.Consumer;
