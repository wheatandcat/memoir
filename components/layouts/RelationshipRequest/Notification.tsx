import Image from "@/components/elements/Image";
import Text from "@/components/elements/Text";
import View from "@/components/elements/View";
import theme from "@/config/theme";
import { MaterialIcons } from "@expo/vector-icons";
import { type FC, memo } from "react";
import { StyleSheet, TouchableOpacity } from "react-native";

export type Props = {
  count: number;
  onPress: () => void;
};

const Notification: FC<Props> = (props) => {
  return (
    <TouchableOpacity onPress={props.onPress}>
      <View style={styles.root}>
        <View>
          <Image
            source={require("@/assets/img/icon/icon_edit.png")}
            width={35}
            height={35}
            contentFit="contain"
          />
        </View>
        <View>
          <Text size="sm">招待申請が{props.count}件届いています</Text>
        </View>
        <View>
          <MaterialIcons name="chevron-right" size={30} />
        </View>
      </View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  root: {
    backgroundColor: theme().color.primary.light,
    width: "100%",
    height: 60,
    alignItems: "center",
    justifyContent: "space-between",
    flexDirection: "row",
    paddingHorizontal: theme().space(2),
  },
});

export default memo(Notification);
