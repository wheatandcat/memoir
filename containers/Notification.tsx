import theme from '@/config/theme';
import { useMutation } from '@apollo/client';
import * as Device from 'expo-device';
import * as Notifications from 'expo-notifications';
import {
  CreatePushTokenDocument,
  type CreatePushTokenMutationVariables,
} from 'queries/api/index';
import type React from 'react';
import {
  createContext,
  memo,
  useCallback,
  useContext,
  useRef,
} from 'react';
import { Platform } from 'react-native';

const Context = createContext<ContextProps>({});
const { Provider } = Context;

type ContextProps = Partial<{
  onPermissionRequest: (callback: () => void) => Promise<boolean>;
}>;

type Props = {
  children: React.ReactNode;
};

Notifications.setNotificationHandler({
  handleNotification: async () => ({
    shouldShowAlert: true,
    shouldPlaySound: false,
    shouldSetBadge: false,
  }),
});

const Notification: React.FC<Props> = memo((props) => {
  const requestCallback = useRef<() => void>(() => {});

  const [createPushTokenMutation] = useMutation(CreatePushTokenDocument, {
    onCompleted() {
      requestCallback.current();
    },
  });

  const onPermissionRequest = useCallback(
    async (callback: () => void) => {
      requestCallback.current = callback;

      let token;

      if (!Device.isDevice) {
        //Alert.alert('端末から実行してくだださい');
        await requestCallback.current();
        return true;
      }

      const { status: existingStatus } =
        await Notifications.getPermissionsAsync();
      let finalStatus = existingStatus;

      if (existingStatus !== 'granted') {
        const { status } = await Notifications.requestPermissionsAsync();
        finalStatus = status;
      }

      if (finalStatus !== 'granted') {
        await requestCallback.current();
        return false;
      }

      token = (await Notifications.getExpoPushTokenAsync()).data;
      console.log(token);

      if (Platform.OS === 'android') {
        Notifications.setNotificationChannelAsync('default', {
          name: 'default',
          importance: Notifications.AndroidImportance.MAX,
          vibrationPattern: [0, 250, 250, 250],
          lightColor: theme().color.primary.main,
        });
      }

      const variables: CreatePushTokenMutationVariables = {
        input: {
          token,
          deviceID: `${(Device.modelName || '').split(' ').join('')}-${String(
            Device.osInternalBuildId || ''
          )
            .split(' ')
            .join('')}`,
        },
      };

      createPushTokenMutation({
        variables,
      });

      return true;
    },
    [createPushTokenMutation]
  );

  return <Provider value={{ onPermissionRequest }}>{props.children}</Provider>;
});

export default Notification;

export const useNotification = () => useContext(Context);
