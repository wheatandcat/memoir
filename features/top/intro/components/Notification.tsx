import Button from "@/components/elements/Button";
import Text from "@/components/elements/Text";
import View from "@/components/elements/View";
import Input from "@/components/layouts/Setting/Memoir/Input";
import NotificationToggle from "@/components/layouts/Setting/Memoir/Notification";
import theme from "@/config/theme";
import type { ConnectedType } from "@/features/top/intro/components/type";
import dayjs from "@/lib/dayjs";
import { StatusBar } from "expo-status-bar";
import { memo, useState } from "react";
import { StyleSheet } from "react-native";
import { SafeAreaView } from "react-native-safe-area-context";

export type Props = Omit<ConnectedType, "onFinish"> & {
  dayOfWeek: number;
  hours: number;
  minutes: number;
  notification: boolean;
  onNext: () => void;
};

const Notification: React.FC<Props> = (props) => {
  const [dayOfWeek, setDayOfWeek] = useState(props.dayOfWeek);
  const [time, setTime] = useState(
    dayjs(
      `0000-01-01T${`${props.hours}`.padStart(
        2,
        "0",
      )}:${`${props.minutes}`.padStart(2, "0")}:00`,
    ).toDate(),
  );
  const [push, setPush] = useState(props.notification ? 1 : 0);

  return (
    <SafeAreaView style={styles.safeArea}>
      <StatusBar style="dark" />
      <View style={styles.header}>
        <Text textAlign="center">初期設定</Text>
      </View>
      <View style={styles.root}>
        <View style={styles.inner}>
          <View>
            <View>
              <Text textAlign="center">いつ、ふりかえりますか？</Text>
            </View>
            <Input
              dayOfWeek={dayOfWeek}
              time={time}
              onChangeDayOfWeek={setDayOfWeek}
              onChangeTime={setTime}
            />
          </View>
          <NotificationToggle push={push} setPush={setPush} />
          <View p={3} style={styles.action}>
            <Button
              title="次へ"
              onPress={() =>
                props.onSaveNotification(
                  {
                    dayOfWeek,
                    hours: dayjs(time).hour(),
                    minutes: dayjs(time).minute(),
                    notification: push === 1,
                  },
                  props.onNext,
                )
              }
              width={200}
            />
          </View>
        </View>
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  safeArea: {
    backgroundColor: theme().color.primary.main,
  },
  root: {
    backgroundColor: theme().color.background.main,
    height: "100%",
    alignItems: "center",
  },
  header: {
    backgroundColor: theme().color.primary.main,
    height: 50,
    alignItems: "center",
    justifyContent: "center",
  },
  inner: {
    paddingVertical: theme().space(3),
  },
  action: {
    alignItems: "center",
    justifyContent: "center",
    marginTop: theme().space(6),
    marginBottom: theme().space(3),
  },
});

export default memo(Notification);
