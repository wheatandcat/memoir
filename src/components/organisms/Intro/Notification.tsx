import React, { memo, useState } from 'react';
import { StyleSheet } from 'react-native';
import { StatusBar } from 'expo-status-bar';
import { SafeAreaView } from 'react-native-safe-area-context';
import View from 'components/atoms/View';
import Text from 'components/atoms/Text';
import theme from 'config/theme';
import Input from 'components/organisms/Setting/Memoir/Input';
import NotificationToggle from 'components/organisms/Setting/Memoir/Notification';
import { ConnectedType } from 'components/pages/Intro/Intro/Connected';
import Button from 'components/atoms/Button';
import dayjs from 'lib/dayjs';

export type Props = ConnectedType & {
  dayOfWeek: number;
  hours: number;
  minutes: number;
  notification: boolean;
  onNext: () => void;
};

const Notification: React.FC<Props> = (props) => {
  const [dayOfWeek, setDayOfWeek] = useState(props.dayOfWeek);
  const [time, setTime] = useState(
    dayjs(
      `0000-01-01T${('00' + props.hours).slice(-2)}:${(
        '00' + props.minutes
      ).slice(-2)}:00`
    ).toDate()
  );
  const [push, setPush] = useState(props.notification ? 1 : 0);

  return (
    <SafeAreaView style={styles.safeArea}>
      <StatusBar backgroundColor={theme().color.primary.main} style="dark" />
      <View style={styles.header}>
        <Text textAlign="center">初期設定</Text>
      </View>
      <View style={styles.root}>
        <View style={styles.inner}>
          <View>
            <View>
              <Text textAlign="center">いつ、ふりかえりますか？</Text>
            </View>
            <Input
              dayOfWeek={dayOfWeek}
              time={time}
              onChangeDayOfWeek={setDayOfWeek}
              onChangeTime={setTime}
            />
          </View>
          <NotificationToggle push={push} setPush={setPush} />
          <View p={3} style={styles.action}>
            <Button
              title="次へ"
              onPress={() =>
                props.onSaveNotification(
                  {
                    dayOfWeek,
                    hours: dayjs(time).hour(),
                    minutes: dayjs(time).minute(),
                    notification: push ? true : false,
                  },
                  props.onNext
                )
              }
              width={200}
            />
          </View>
        </View>
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  safeArea: {
    backgroundColor: theme().color.primary.main,
  },
  root: {
    backgroundColor: theme().color.background.main,
    height: '100%',
    alignItems: 'center',
  },
  header: {
    backgroundColor: theme().color.primary.main,
    height: 50,
    alignItems: 'center',
    justifyContent: 'center',
  },
  inner: {
    paddingVertical: theme().space(3),
  },
  action: {
    alignItems: 'center',
    justifyContent: 'center',
    marginTop: theme().space(6),
    marginBottom: theme().space(3),
  },
});

export default memo(Notification);
