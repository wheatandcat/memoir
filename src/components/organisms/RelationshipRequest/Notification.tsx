import React, { memo } from 'react';
import { StyleSheet, TouchableOpacity } from 'react-native';
import View from 'components/atoms/View';
import Text from 'components/atoms/Text';
import Image from 'components/atoms/Image';
import theme from 'config/theme';
import { MaterialIcons } from '@expo/vector-icons';

export type Props = {
  count: number;
  onPress: () => void;
};

const Notification: React.FC<Props> = (props) => {
  return (
    <TouchableOpacity onPress={props.onPress}>
      <View style={styles.root}>
        <View>
          <Image
            source={require('../../../img/icon/icon_edit.png')}
            width={35}
            height={35}
            contentFit="contain"
          />
        </View>
        <View>
          <Text size="sm">招待申請が{props.count}件届いています</Text>
        </View>
        <View>
          <MaterialIcons name="chevron-right" size={30} />
        </View>
      </View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  root: {
    backgroundColor: theme().color.primary.light,
    width: '100%',
    height: 60,
    alignItems: 'center',
    justifyContent: 'space-between',
    flexDirection: 'row',
    paddingHorizontal: theme().space(2),
  },
});

export default memo(Notification);
