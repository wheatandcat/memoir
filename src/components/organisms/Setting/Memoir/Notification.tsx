import React, { memo } from 'react';
import { StyleSheet } from 'react-native';
import SwitchSelector from 'react-native-switch-selector';
import View from 'components/atoms/View';
import Text from 'components/atoms/Text';
import theme from 'config/theme';

export type Props = {
  push: number;
  setPush: (push: number) => void;
};

const options = [
  { label: 'OFF', value: 0 },
  { label: 'ON', value: 1 },
];

const Notification: React.FC<Props> = (props) => {
  return (
    <>
      <View style={styles.title}>
        <View style={styles.line} />

        <View>
          <Text>通知設定</Text>
        </View>

        <View style={styles.line} />
      </View>
      <View style={styles.push}>
        <View>
          <Text variants="small">プッシュ通知</Text>
        </View>
        <View>
          <SwitchSelector
            initial={props.push}
            options={options}
            bold
            buttonColor={theme().color.primary.main}
            backgroundColor={theme().color.base.main}
            fontSize={theme().fontSizes.base}
            textColor={theme().color.base.light}
            selectedColor={theme().color.secondary.main}
            style={styles.pushInput}
            height={43}
            onPress={(v) => props.setPush(Number(v))}
          />
        </View>
      </View>
    </>
  );
};

const styles = StyleSheet.create({
  title: {
    marginTop: theme().space(5),
    flexDirection: 'row',
    height: 50,
    alignItems: 'center',
    justifyContent: 'space-around',
  },
  line: {
    width: '30%',
    backgroundColor: theme().color.base.dark,
    height: StyleSheet.hairlineWidth,
    marginHorizontal: theme().space(3),
  },
  push: {
    marginTop: theme().space(4),
    paddingHorizontal: theme().space(4),
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  pushInput: {
    width: 139,
  },
});

export default memo(Notification);
