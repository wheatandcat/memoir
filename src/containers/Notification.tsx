import * as Device from 'expo-device';
import * as Notifications from 'expo-notifications';
import React, {
  memo,
  useRef,
  useContext,
  useCallback,
  createContext,
} from 'react';
import { Platform } from 'react-native';
import {
  useCreatePushTokenMutation,
  CreatePushTokenMutationVariables,
} from 'queries/api/index';
import theme from 'config/theme';

const Context = createContext<ContextProps>({});
const { Provider } = Context;

type ContextProps = Partial<{
  onPermissionRequest: (callback: () => void) => Promise<boolean>;
}>;

type Props = {};

Notifications.setNotificationHandler({
  handleNotification: async () => ({
    shouldShowAlert: true,
    shouldPlaySound: false,
    shouldSetBadge: false,
  }),
});

const Notification: React.FC<Props> = memo((props) => {
  const requestCallback = useRef<() => void>(() => {});

  const [createPushTokenMutation] = useCreatePushTokenMutation({
    onCompleted() {
      requestCallback.current();
    },
  });

  const onPermissionRequest = useCallback(
    async (callback: () => void) => {
      requestCallback.current = callback;

      let token;

      if (!Device.isDevice) {
        //Alert.alert('端末から実行してくだださい');
        requestCallback.current();
        return true;
      }

      const { status: existingStatus } =
        await Notifications.getPermissionsAsync();
      let finalStatus = existingStatus;

      if (existingStatus !== 'granted') {
        const { status } = await Notifications.requestPermissionsAsync();
        finalStatus = status;
      }

      if (finalStatus !== 'granted') {
        requestCallback.current();
        return false;
      }

      token = (await Notifications.getExpoPushTokenAsync()).data;
      console.log(token);

      if (Platform.OS === 'android') {
        Notifications.setNotificationChannelAsync('default', {
          name: 'default',
          importance: Notifications.AndroidImportance.MAX,
          vibrationPattern: [0, 250, 250, 250],
          lightColor: theme().color.primary.main,
        });
      }

      const variables: CreatePushTokenMutationVariables = {
        input: {
          token,
          deviceID: `${(Device.modelName || '').split(' ').join('')}-${String(
            Device.osInternalBuildId || ''
          )
            .split(' ')
            .join('')}`,
        },
      };

      createPushTokenMutation({
        variables,
      });

      return true;
    },
    [createPushTokenMutation]
  );

  return <Provider value={{ onPermissionRequest }}>{props.children}</Provider>;
});

export default Notification;

export const useNotification = () => useContext(Context);
