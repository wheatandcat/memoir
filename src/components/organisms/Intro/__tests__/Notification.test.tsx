import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import Notification, { Props } from '../Notification';

const propsData = (): Props => ({});

describe('components/organisms/Intro/Notification.tsx', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<Notification {...propsData()} />);
  });

  it('正常にrenderすること', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
