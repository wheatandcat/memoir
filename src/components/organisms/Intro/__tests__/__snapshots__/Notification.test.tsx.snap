// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/organisms/Intro/Notification.tsx 正常にrenderすること 1`] = `
<ForwardRef
  style={
    Object {
      "backgroundColor": "#E3C95D",
    }
  }
>
  <ExpoStatusBar
    backgroundColor="#E3C95D"
    style="dark"
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#E3C95D",
        "height": 50,
        "justifyContent": "center",
      }
    }
  >
    <Text
      fontFamily="RobotoCondensed-Bold"
      textAlign="center"
      variants="body"
    >
      初期設定
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F2F2F2",
        "height": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "paddingVertical": 16,
        }
      }
    >
      <View>
        <View>
          <Text
            fontFamily="RobotoCondensed-Bold"
            textAlign="center"
            variants="body"
          >
            いつ、ふりかえりますか？
          </Text>
        </View>
        <Memo(Input)
          dayOfWeek={1}
          onChangeDayOfWeek={[Function]}
          onChangeTime={[Function]}
          time={1899-12-31T15:00:00.000Z}
        />
      </View>
      <Memo(Notification)
        push={0}
        setPush={[Function]}
      />
      <View
        p={3}
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "marginBottom": 16,
            "marginTop": 128,
          }
        }
      >
        <Button
          fontFamily="NotoSansJP-Bold"
          onPress={[Function]}
          size="base"
          title="次へ"
          width={200}
        />
      </View>
    </View>
  </View>
</ForwardRef>
`;
